OADP-6673: Add parallel backup (#111001)
openshift/openshift-docs: backup_and_restore/application_backup_and_restore/backing_up_and_restoring/oadp-parallel-backup-processing-in-oadp.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc
include::_attributes/common-attributes.adoc[]

:_mod-docs-content-type: CONCEPT
[id="oadp-parallel-backup-processing-in-oadp_{context}"]
= Parallel backup processing in {oadp-short}

[role="_abstract"]
With parallel backup processing in {oadp-full}, you can back up several applications from different namespaces simultaneously. This concurrent processing ensures that a single large backup task does not delay the backups for other applications in your cluster.

> 

Reviewer: The short desc is a lil over 300 characters (303). It will get flagged in the CQA report. How about this:

`Back up several applications from different namespaces simultaneously with parallel backup processing in {oadp-full}. This helps you prevent a single large backup from delaying backups for other applications in your cluster.`

Author: How are you counting characters @shdeshpa07 ?

Reviewer: Claude Code ;)..definitely not me.

Author: I double checked and is getting 282 characters so I'll keep the existing one.

By default, Velero processes backup resources sequentially - one resource in the `InProgress` phase at a time. Parallel backup changes this behavior to handle backups concurrently. 

> The following considerations apply for enabling parallel backups in OADP:

Reviewer: ```suggestion
The following considerations apply for enabling parallel backups in {oadp-short}:
```

Author: Ah this one is embarrassing. Thank you!

* Resource allocation: Increasing the concurrency settings requires additional CPU and memory for the Velero container. Monitor resource consumption to ensure cluster stability.

* Backup scope: Parallel processing provides the most benefit for backups that contain a large number of Kubernetes resources or many smaller volumes. Backups dominated by a few large volumes see less benefit because most time is spent waiting for asynchronous data movement to complete.

> * Namespace partitioning: To keep data integrity, {oadp-short} uses namespace-level isolation when processing parallel backups. Two backups cannot run simultaneously if they share any included namespaces.

Reviewer 2: ```suggestion
* Namespace partitioning: To keep data integrity, {oadp-short} uses namespace-level isolation when processing parallel backups. Two backups cannot run simultaneously if any resources are shared or overlap between namespaces.
```
@sseago do you agree w/ my suggestion? Perhaps have a better way to express it?   Your original text..   ```
By default, only one backup is processed in the `InProgress` phase at a time. The install flag `concurrent-backups`, which takes an integer argument, configures Velero to process multiple backups at the same time, up to a max of `concurrent-backups`. The other restriction on parallel backup processing is that two backups which have any included namespaces in common may not run at the same time. For example, if `concurrent-backups` is set to 2 and two backups for "namespace1" are submitted at the same time, only one of those will be processed at the same time. On the other hand, if a backup for "namespace1" and another for "namespace2" are submitted, then both can be processed in parallel. Note that a whole-cluster backup (one which does not restrict to a set list of namespaces) includes all namespaces, and therefore it will not run in parallel with any other backup.
```

Reviewer 3: @weshayutin I think referencing shared namespaces is clearer than shared resources, since that directly references the included/excluded namespaces fields in the backup. If you reference resources instead of namespaces, users might get confused and think there should be no overlap in the `includedResources` field, which is not true at all. Also, the "any resources overlap" is not strictly correct, since we only calculate overlap at the namespace level. If a cluster-scoped resource is included in 2 backups at the same time, that won't prevent scheduling.

Author: No changes will be made, keeping namespaces. Thank you @weshayutin and @sseago for your feedback!
